### ISSUE TITLE
A subquery result is different from PostgreSQL

### ISSUE THREAD

**Issue by chenzl25** [risingwavelabs/risingwave#4203]
```sql
create table t1 (x int, y int);
create table t2 (x int, y int);
create table t3(x int, y int);
insert into t1 values (1, 1), (2, 2), (1, null), (null, 2), (2, null), (null, 1), (null, null);
insert into t2 values (1, 2), (2, 2), (1, null), (null, 2), (2, null), (null, 1), (null, null);
insert into t3 values (1, 2), (2, 2), (1, null), (null, 2), (2, null), (null, 1), (null, null);

select * from t1 where exists(select t2.x from t2 left join t3 on t2.x = t3.x and t1.y = t2.y and t1.y = t3.y);

risingwave result:
 x | y
---+---
   | 1
   | 2
 2 | 2
 1 | 1
(4 rows)

postgresql result:
 x | y
---+---
 1 | 1
 2 | 2
 1 |
   | 2
 2 |
   | 1
   |
(7 rows)
```

**Comment by chenzl25** [comment]
```
 Begin:

 LogicalProject { exprs: [t1.x, t1.y] }
   LogicalApply { type: LeftSemi, on: true, correlated_id: 1 }
     LogicalScan { table: t1, columns: [t1._row_id, t1.x, t1.y] }
     LogicalProject { exprs: [t2.x] }
       LogicalJoin { type: LeftOuter, on: (t2.x = t3.x) AND (CorrelatedInputRef { index: 2, correlated_id: 1 } = t2.y) AND (CorrelatedInputRef { index: 2, correlated_id: 1 } = t3.y), output_indices: all }
         LogicalScan { table: t2, columns: [t2._row_id, t2.x, t2.y] }
         LogicalScan { table: t3, columns: [t3._row_id, t3.x, t3.y] }

 General Unnesting(Translate Apply):

 apply TranslateApplyRule 1 time(s)

 LogicalProject { exprs: [t1.x, t1.y] }
   LogicalJoin { type: LeftSemi, on: (t1.y = t1.y), output_indices: all }
     LogicalScan { table: t1, columns: [t1._row_id, t1.x, t1.y] }
     LogicalApply { type: Inner, on: true, correlated_id: 1 }
       LogicalAgg { group_key: [t1.y], aggs: [] }
         LogicalProject { exprs: [t1.y] }
           LogicalScan { table: t1, columns: [t1.y] }
       LogicalProject { exprs: [t2.x] }
         LogicalJoin { type: LeftOuter, on: (t2.x = t3.x) AND (CorrelatedInputRef { index: 2, correlated_id: 1 } = t2.y) AND (CorrelatedInputRef { index: 2, correlated_id: 1 } = t3.y), output_indices: all }
           LogicalScan { table: t2, columns: [t2._row_id, t2.x, t2.y] }
           LogicalScan { table: t3, columns: [t3._row_id, t3.x, t3.y] }
```

TranslateApplyRule add a join on top of apply with condition like t1.y = t1.y which is null rejected. So, if the original Apply with a condition true(that is the exists case) all the null row would be filter out. I think a `is distinct from` instead of `equal` semantic HashJoin could help.

**Comment by chenzl25** [comment]
Fixed By #5021
